godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,875 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::cell::RefCell;
+
+use crate::engine::{Object, Script, ScriptLanguage};
+use crate::obj::Gd;
+use crate::sys;
+
+use super::meta::{MethodInfo, PropertyInfo};
+use super::{GString, StringName, Variant, VariantType};
+
+/// Interface for Godot's `GDExtensionScriptInstance`, which cannot be directly constructed by an extension. Instead a type that implements this
+/// trait has to be passed to the [`create_script_instance`] function, which returns a [`sys::GDExtensionScriptInstancePtr`] pointer. This pointer
+/// can then be returned from [`IScriptExtension::instance_create`](crate::engine::IScriptExtension::instance_create).
+pub trait ScriptInstance {
+    /// Name of the new class the script implements.
+    fn class_name(&self) -> GString;
+
+    /// Property setter for Godots virtual dispatch system. The engine will call this function when it wants to change a property on the script.
+    fn set(&mut self, name: StringName, value: &Variant) -> bool;
+
+    /// Property getter for Godots virtual dispatch system. The engine will call this function when it wants to read a property on the script.
+    fn get(&self, name: StringName) -> Option<Variant>;
+
+    /// A list of all the properties a script exposes to the engine.
+    fn property_list(&self) -> &[PropertyInfo];
+
+    /// A list of all the methods a script exposes to the engine.
+    fn method_list(&self) -> &[MethodInfo];
```

Comment:
Also here, `get_` prefix would be consistent with [`Object::get_{property|method}_list`](https://godot-rust.github.io/docs/gdext/master/godot/engine/struct.Object.html#method.get_property_list).